src/ast.rs

```diff
@@ -3891,11 +3298,11 @@
     if top_type == Type::job_list {
         let mut list = pops.allocate::<JobList>();
         pops.populate_list(&mut *list, true /* exhaust_stream */);
-        ast.top = list;
+        ast.top = Box::new(NodeEnum::from(*list));
```

Comment:
> I tried to avoid introducing custom proc macros, but a mysterious hygiene
> bug forced my hand.

I get `error[E0412]: cannot find type `BranchEnum` in this scope`
Can't that be fixed?

> Plus, it's arguably more readable than the declarative version, and the
> blanket_ref attribute is a nice bonus (those blanket impls are necessary
> for enum_dispatch to do its thing on the ref enums)

I wonder how much we'd miss if we implemented enum_dispatch ourselves.
I have a preference for avoiding proc macros -- it should be possible
to implement a light version of enum_dipatch with normal macros,
maybe not match statements but if-else chains -- but I agree the
`enum` syntax works great. So this is probably the best way forward.

Reply:
> I get "error[E0412]: cannot find type `BranchEnum` in this scope"
Not sure what you mean by that. On 1.77.0, the MRE I linked produces "cannot find value `inner` in this scope".
xiretza found [this old PR](https://gitlab.com/antonok/enum_dispatch/-/merge_requests/22) that allegedly fixeds the problem.
However, if we went back to a declarative macro, we'd need to use `paste`, and when I tried using that here, the readability was terrible.

There is a possible future where the design simplifies to the point that we can just have the enums defined without a macro, but we need to get rid of the `Enum` and `RefMut` flavors first to avoid descending (deeper?) into boilerplate hell.